bootstrap/cmd/bootstrap/app/kfctlUtil.go

```diff
@@ -0,0 +1,112 @@
+package app
+
+import (
+	"encoding/base64"
+	kftypes "github.com/kubeflow/kubeflow/bootstrap/pkg/apis/apps"
+	kfdefs "github.com/kubeflow/kubeflow/bootstrap/pkg/apis/apps/kfdef/v1alpha1"
+	log "github.com/sirupsen/logrus"
+	"github.com/kubeflow/kubeflow/bootstrap/pkg/kfapp/gcp"
+	"github.com/kubeflow/kubeflow/bootstrap/pkg/kfapp/ksonnet"
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"os"
+	"path"
+)
+
+// ToKfdef will output CreateRequest in format of kfdefs.KfDef
+func (cr *CreateRequest) ToKfdef(appDir string, repo string, istio bool) (*kfdefs.KfDef, error) {
+	kfDef := &kfdefs.KfDef{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "KfDef",
+			APIVersion: "kfdef.apps.kubeflow.org/v1alpha1",
+		},
+		Spec: kfdefs.KfDefSpec{},
+	}
+	kfDef.Spec.ComponentConfig = cr.AppConfig
+	kfDef.Name = cr.Name
+	kfDef.Spec.AppDir = appDir
+	kfDef.Spec.Email = cr.Email
+	kfDef.Spec.Platform = kftypes.GCP
+	kfDef.Namespace = cr.Namespace
+	kfDef.Spec.Version = cr.KfVersion
+	kfDef.Spec.Repo = repo
+	kfDef.Spec.Project = cr.Project
+	kfDef.Spec.IpName = cr.IPName
+	kfDef.Spec.Zone = cr.Zone
+	if cr.Username != "" {
+		kfDef.Spec.UseBasicAuth = true
+	}
+	kfDef.Spec.SkipInitProject = true
+	kfDef.Spec.UseIstio = istio
+	return kfDef, nil
+}
+
+func (s *ksServer) DeployWithKfctl(req *CreateRequest) error{
+	// pull versioned kubeflow repo
+	ksRegistry := kfdefs.GetDefaultRegistry()
+	ksRegistry.Version = req.KfVersion
+	versionedRegPath, err := s.getRegistryUri(ksRegistry)
+	if err != nil {
+		return err
+	}
+
+	// pull source repo to random tmp dir
+	repoDir, err := s.CloneRepoToLocal(req.Project, req.Token, GetRepoNameKfctl(req.Project))
+	// clean up tmp dir afterwards
+	defer os.RemoveAll(repoDir)
+	if err != nil {
+		return err
+	}
+	ksAppDir := path.Join(repoDir, GetRepoNameKfctl(req.Project), req.KfVersion, req.Name)
+	kfdef, err := req.ToKfdef(ksAppDir, versionedRegPath, s.installIstio)
+	if err != nil {
+		return err
+	}
+	ts := oauth2.StaticTokenSource(&oauth2.Token{
+		AccessToken: req.Token,
+	})
+	ctx := context.Background()
+
+	var gcpApp kftypes.KfApp
+	// run gcp generate / apply
+	if kfdef.Spec.UseBasicAuth {
+		UsernameData, err := base64.StdEncoding.DecodeString(req.Username)
+		if err != nil {
+			log.Errorf("Failed decoding username: %v", err)
+			return err
+		}
+		gcpApp = gcp.BuildKfApp(kfdef, oauth2.NewClient(ctx, ts), ts, req.StorageOption, req.SAClientID,
+			string(UsernameData), req.PasswordHash, "", "")
+	} else {
+		gcpApp = gcp.BuildKfApp(kfdef, oauth2.NewClient(ctx, ts), ts, req.StorageOption, req.SAClientID,
+			"", "", req.ClientID, req.ClientSecret)
+	}
+	
+	if err = gcpApp.Generate(kftypes.ALL); err != nil {
```

Comment:
is it required to do gcp.Generate, Apply then ksonnet.Generate, Apply?
If coordinator is used it would be coordinator calling gcp.Generate, ksonnet/kustomize.Generate, gcp.Apply, ksonnet/kustomize.Apply